design/healcheck-endpoint.md

```diff
@@ -0,0 +1,159 @@
+and External IdPs probes are free of charge as we report success/failure calls for them in the Metrics components.
+With this approach, the Healthcheck Endpoint becomes an alternative reporting tool for the Metrics.
+
+This design assumes implementing Health Endpoint in Passive Mode only.
+
+#### Microprofile Health
+
+[Eclipse Microprofile Health](https://download.eclipse.org/microprofile/microprofile-health-3.0-RC4/microprofile-health-spec.html)
+project provides capabilities for reporting Liveness and Readiness probes. In the contrast to Microprofile Metrics,
+the Health project doesn't provide a centralized class to access all configured checks. The Smallrye implementation
+provides `SmallRyeHealthReporter` (see [the code](https://github.com/smallrye/smallrye-health/blob/73e9ea8c31498dab7753f4a428cb424bf9fbaa1a/implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java#L49))
+, that is a CDI-annotated Bean that contains references to `HealthCheck` implementations. The `HealthCheck` interface
+implementations provides a specific check - e.g. check if a database is up.
+
+Keycloak will provide its own implementation of a class similar to `SmallRyeHealthReporter` - `KeycloakHealthReporter`.
+Since `SmallRyeHealthReporter` is relies heavily on CDI (it uses `Instance<Healthcheck>` for injection), we need to
+provide a class with similar capabilities that enables us to assemble a centralized Health reporting instance manually.
+
+Individual Healthcheck implementations, such as database check, External IdP check or LDAP check need to be provided
+as `HealthCheck` class implementations. Later on, they will be added to `KeycloakHealthReporter`.
+
+#### Keycloak REST integration
+
+Integrating with Wildfly's `/health` endpoint requires either adding CDI annotations to Keycloak and relying on CDI
+deployment processor (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/deployment/CDIExtension.java))
+or integrating with Microprofile Health Subsystem. An additional problem is that Wildfly provides its own
+`HealthReporter` class (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/HealthReporter.java))
+that doesn't use groups. The Grouping functionality is essential for this design (explained below).
+
+The biggest downside of providing Keycloak's own implementation is that we'll end up with 2 health endpoints -
+one for Wildfly (`http://localhost:9990/health`) and one for Keycloak (`https://localhost:8443/auth/health`). We
+need to be very clear in the documentation that the latter is the proper one.
+
+Having the above in mind, the design assumes integrating the new Healthcheck API with Keycloak's REST interface.
+The implementation will create additional `/health` endpoint that will parse incoming query parameters. Once all
```

Comment:
There should obviously only be one health endpoint, so if we have our own we'd need to also run any relevant underlying container health checks. What's the overhead of doing that?

Reply:
Wildfly's Healthcheck Endpoint is focused on the server and deployed applications. It checks the following things (see [the code](https://github.com/wildfly/wildfly/blob/cf2ecf8cd575f425ba5cb254023fd6d44f2aa070/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/ServerReadinessProbes.java)):
- The `server-state` attribute.
- Checks the `read-boot-errors` from Core Service
- Checks if all deployments are fine

If the new Healthcheck endpoint is incorporated into Keycloak's codebase, none of the above checks applies. If anything bad happens during the boot time, Keycloak's JAX-RS endpoints will be down and the probe will naturally fail. 

The only questionable thing is checking the deployed Providers. But I'm not sure how to do it properly. Perhaps we could implement an additional check:

    curl https://<keycloak-url>:8443/auth/health?fail-on=provider-my-provider-id

But I'm not keen of doing that. I've always perceived deploying custom Providers to Keycloak as a configuration thing.

Reply:
KC start should probably really just fail if a custom provider fails, as we can't expect KC to operate in a healthy way without an expected provider. So we don't need a health check for providers as such.

Reply:
I agree.

Reply:
I believe this conversation can be resolved.